File: web-ts/src/state/hubTrustProvenance.test.ts
import { describe, it, expect } from 'vitest'
import { emptyTrustProvenance, recordObservation, displayedTrust } from './hubTrustProvenance'

describe('CYP-853 (Multi-Hub M3) — hubTrustProvenance: switch-first trust degradation (MC-2, immediate lapse)', () => {
  it('the ACTIVE hub shows its live observed trust (we are observing it now)', () => {
    const p = recordObservation(emptyTrustProvenance, 'local', 'TRUSTED', 1000)
    expect(displayedTrust(p, 'local', 'local')).toBe('TRUSTED')
  })

  it('the ACTIVE hub is UNKNOWN until first observed (fail-closed, never optimistic)', () => {
    expect(displayedTrust(emptyTrustProvenance, 'local', 'local')).toBe('UNKNOWN')
  })

  it('★ an INACTIVE hub that WAS trusted → STALE, NEVER cached-`trusted` (the core honesty bug this forbids)', () => {
    // MUT: displayedTrust returning obs.trust for the inactive branch (keep showing cached trusted) → reds.
    const p = recordObservation(emptyTrustProvenance, 'alpha', 'TRUSTED', 1000)
    expect(displayedTrust(p, 'alpha', 'local')).toBe('STALE')
  })

  it('★ an INACTIVE hub NEVER observed → UNKNOWN, never a fabricated STALE (stale ⇒ there WAS an observation)', () => {
    // MUT: never-observed → STALE (inventing a prior observation) → reds.
    expect(displayedTrust(emptyTrustProvenance, 'alpha', 'local')).toBe('UNKNOWN')
  })

  it('★ an INACTIVE hub that was REJECTED → REJECTED (a fail-closed negative is preserved, not softened)', () => {
    const p = recordObservation(emptyTrustProvenance, 'alpha', 'REJECTED', 1000)
    expect(displayedTrust(p, 'alpha', 'local')).toBe('REJECTED')
  })

  it('an INACTIVE hub last seen pending/unknown → UNKNOWN (nothing affirmative to show without observation)', () => {
    const pPending = recordObservation(emptyTrustProvenance, 'alpha', 'PENDING', 1000)
    expect(displayedTrust(pPending, 'alpha', 'local')).toBe('UNKNOWN')
  })

  it('★ MC-1 no cross-hub leak: hub A’s observed trust does NOT bleed into hub B’s display', () => {
    // A observed trusted, B never observed; active = A.
    const p = recordObservation(emptyTrustProvenance, 'A', 'TRUSTED', 1000)
    expect(displayedTrust(p, 'A', 'A')).toBe('TRUSTED') // active A: live
    expect(displayedTrust(p, 'B', 'A')).toBe('UNKNOWN') // B never observed — A's trust must NOT leak to B
    // switch active to C: A lapses to STALE (its OWN provenance), B still UNKNOWN (unchanged) — per-hub, isolated.
    expect(displayedTrust(p, 'A', 'C')).toBe('STALE')
    expect(displayedTrust(p, 'B', 'C')).toBe('UNKNOWN')
  })

  it('★ provenance carries the observation TIMESTAMP (the seam for a future window), and a later observation replaces it', () => {
    let p = recordObservation(emptyTrustProvenance, 'alpha', 'TRUSTED', 1000)
    expect(p['alpha']).toEqual({ trust: 'TRUSTED', observedAt: 1000 })
    p = recordObservation(p, 'alpha', 'REJECTED', 2000) // a later observation supersedes
    expect(p['alpha']).toEqual({ trust: 'REJECTED', observedAt: 2000 })
    expect(displayedTrust(p, 'alpha', 'alpha')).toBe('REJECTED')
  })
})

File: app/shared/src/commonTest/kotlin/com/tneff/cyppieagents/multihub/HubTrustProvenanceTest.kt
package com.tneff.cyppieagents.multihub

import com.tneff.cyppieagents.model.HubTrustState
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

/**
 * CYP-865 — the switch-first trust-degradation honesty matrix ([displayedTrust]): the ACTIVE hub shows its live
 * observed trust (UNKNOWN until observed); an INACTIVE hub is not observed → last-known lapses (was-TRUSTED → STALE,
 * NEVER cached-trusted; was-REJECTED preserved; anything else → UNKNOWN). Fail-closed throughout.
 */
class HubTrustProvenanceTest {

    @Test
    fun active_liveObserved_returnsTheObservedTrust() {
        val p = recordObservation(emptyTrustProvenance, "a", HubTrustState.TRUSTED, observedAt = 1L)
        assertEquals(HubTrustState.TRUSTED, displayedTrust(p, hubId = "a", activeHubId = "a"))
    }

    @Test
    fun active_neverObserved_isUnknown() {
        assertEquals(HubTrustState.UNKNOWN, displayedTrust(emptyTrustProvenance, hubId = "a", activeHubId = "a"))
    }

    @Test
    fun inactive_wasTrusted_isStale_neverCachedTrusted() {
        // The load-bearing honesty rule: a switched-away hub that WAS trusted lapses to STALE — never keep showing
        // a cached TRUSTED (that would be derivation-masquerading-as-observation). Mutation TRUSTED→TRUSTED → RED.
        val p = recordObservation(emptyTrustProvenance, "b", HubTrustState.TRUSTED, observedAt = 1L)
        assertEquals(HubTrustState.STALE, displayedTrust(p, hubId = "b", activeHubId = "a"))
    }

    @Test
    fun inactive_wasRejected_staysRejected_failClosedNegativePreserved() {
        val p = recordObservation(emptyTrustProvenance, "b", HubTrustState.REJECTED, observedAt = 1L)
        assertEquals(HubTrustState.REJECTED, displayedTrust(p, hubId = "b", activeHubId = "a"))
    }

    @Test
    fun inactive_neverObserved_isUnknown() {
        assertEquals(HubTrustState.UNKNOWN, displayedTrust(emptyTrustProvenance, hubId = "b", activeHubId = "a"))
    }

    @Test
    fun inactive_wasPendingOrUnknownOrStale_isUnknown_nothingFreshToAffirm() {
        for (t in listOf(HubTrustState.PENDING, HubTrustState.UNKNOWN, HubTrustState.STALE)) {
            val p = recordObservation(emptyTrustProvenance, "b", t, observedAt = 1L)
            assertEquals(HubTrustState.UNKNOWN, displayedTrust(p, hubId = "b", activeHubId = "a"), "inactive was $t → UNKNOWN")
        }
    }

    @Test
    fun recordObservation_isPure_immutable_carriesProvenanceTime() {
        val p1 = emptyTrustProvenance
        val p2 = recordObservation(p1, "a", HubTrustState.TRUSTED, observedAt = 42L)
        assertTrue(p1.isEmpty(), "recordObservation is pure — the prior map is untouched")
        assertEquals(HubTrustState.TRUSTED, p2["a"]?.trust)
        assertEquals(42L, p2["a"]?.observedAt, "the observation time is carried as honest provenance")
    }

    @Test
    fun recordObservation_overwrites_withLatestObservation() {
        val p = recordObservation(
            recordObservation(emptyTrustProvenance, "a", HubTrustState.TRUSTED, observedAt = 1L),
            "a", HubTrustState.REJECTED, observedAt = 2L,
        )
        assertEquals(HubTrustState.REJECTED, p["a"]?.trust)
        assertEquals(2L, p["a"]?.observedAt)
    }
}
